### PR TITLE
implement linux autostart

### PR DESCRIPTION
**Pull request by jdheeter**
I'm not able to test locally yet but will try tomorrow.

### CODE REVIEW

File: src/lib/autoLaunch/linuxAutoLaunch.ts [hunk 1]
```diff
@@ -1,29 +1,44 @@
+import { AutoLaunchParams, ChildReturnData } from '../types'
+import * as native from '../native'
+import * as fs from "@tauri-apps/api/fs"
 
-module.exports = {
+const linuxAutoLaunch = {
```
Comment (nazar-pc) [inline src/lib/autoLaunch/linuxAutoLaunch.ts:5]: Can we have this as a class that implements an interface shared for all 3 platforms?

File: src/lib/autoLaunch/linuxAutoLaunch.ts [hunk 2]
```diff
@@ -1,29 +1,44 @@
...
+    try {
+      await fs.readTextFile(this.getFilePath(appName))
+      return true
+    } catch (error) {
+      console.error(error)
+      return false
+    }
+  },
+  getDirectory(): string { return '~/.config/autostart/'; },
+  getFilePath(appName: string): string { return `${this.getDirectory()}${appName}.desktop`; }
```
Comment (nazar-pc) [inline src/lib/autoLaunch/linuxAutoLaunch.ts:42]: I don't think these are supposed to be public API.

File: src/lib/autoLaunch/linuxAutoLaunch.ts [hunk 3]
```diff
@@ -1,29 +1,44 @@
...
-};
+    const contents = `
+    [Desktop Entry]
+    Type=Application
+    Version=1.0
+    Name=${appName}
+    Comment=${appName}startup script
+    Exec=${appPath}${hiddenArg}
+    StartupNotify=false
+    Terminal=false
```
Comment (nazar-pc) [inline src/lib/autoLaunch/linuxAutoLaunch.ts:19]:
```suggestion
[Desktop Entry]
Type=Application
Name=${appName}
Comment=${appName} startup script
Exec=${appPath}${hiddenArg}
Icon=PATH_TO_APP_ICON
```

File: src/lib/autoLaunch/linuxAutoLaunch.ts (hunk 1 above)
Comment (jdheeter) [reply src/lib/autoLaunch/linuxAutoLaunch.ts:5]: The class that implements the cross platform functions is in native.ts but I agree that things could be organized in a better way.

File: src-tauri/src/main.rs [hunk 4]
```diff
@@ -100,6 +100,8 @@ fn main() {
     windows::winreg_get,
     windows::winreg_set,
     windows::winreg_delete,
+    get_this_binary,
```
Comment (jdheeter) [inline src-tauri/src/main.rs:103]: @nazar-pc I found that this command was not compiled on windows unless I add it here.

File: src/lib/autoLaunch/linuxAutoLaunch.ts (hunk 2 above)
Comment (jdheeter) [reply src/lib/autoLaunch/linuxAutoLaunch.ts]: consolidated in latest commit

File: src-tauri/src/main.rs (hunk 4 above)
Comment (nazar-pc) [reply src-tauri/src/main.rs:103]: Yeah, that is correct after I read the implementation in Tauri, thanks!

File: src/lib/autoLauncher.ts [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  }
+  async enable(): Promise<void | ChildReturnData> {
+    const child = await this.autoLauncher.enable({ appName: this.appName, appPath: this.appPath, hidden: false })
+    return child
+  }
+  async disable(): Promise<void | ChildReturnData> {
+    const child = this.autoLauncher.disable(this.appName)
+    return child
+  }
+  async init(): Promise<void> {
```
Comment (nazar-pc) [inline src/lib/autoLauncher.ts:130]: The same flawed approach with initialization here, please either make sure it is impossible to instantiate uninitialized `AutoLauncher` or add TODO to fix that.

File: src/lib/autoLauncher.ts [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    const exists = loginList.includes(appName)
+    console.log('login Item Exists:', exists);
+    return exists
+  },
+
+}
+
+const winAL = {
+  subKey: "SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Run",
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (nazar-pc) [inline src/lib/autoLauncher.ts:88]: Shouldn't you have TODO to support `hidden`? Also if I understand it, better name would be `minimized`.

File: src/lib/autoLauncher.ts [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+    const contents = `
+    [Desktop Entry]
+    Type=Application
+    Version=1.0
+    Name=${appName}
+    Comment=${appName}startup script
+    Exec=${appPath}${hiddenArg}
+    StartupNotify=false
+    Terminal=false
```
Comment (nazar-pc) [inline src/lib/autoLauncher.ts:40]:
```suggestion
[Desktop Entry]
Type=Application
Name=${appName}
Comment=${appName} startup script
Exec=${appPath}${hiddenArg}
Icon=PATH_TO_APP_ICON
```

File: src/lib/autoLauncher.ts [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+    if (result.search('The system cannot find the file specified.') > -1) return false
+    else return true
+  }
+}
+
+
+export class AutoLauncher {
+  protected autoLauncher: osAL = nullAL
+  appName = 'app'
+  appPath = ''
```
Comment (nazar-pc) [inline src/lib/autoLauncher.ts:115]: Should this field be public? If so why?

File: src/lib/autoLauncher.ts (hunk 6 above)
Comment (jdheeter) [reply src/lib/autoLauncher.ts:88]: Minimized implies that there is a window which is minimized to the taskbar but what this actually means is don't spawn a widow at all. In hidden mode the app would just start in the tray on the respective platform with no window would be spawned until you open the tray menu. Added a TODO to handle this case on windows.

Comment (nazar-pc) [reply src/lib/autoLauncher.ts:88]: What describe is that the window is effectively there, it is just not visible until you open the tray menu. I believe it is called `minimized` in most other apps, not `hidden`. In Linux at least there is actually a separate `hidden` attribute in the autostart file which will make it invisible in GUI that manages autostarting apps.

Comment (jdheeter) [reply src/lib/autoLauncher.ts:89]: Ok, I'm not sure of the specifics, this nomenclature came from the npm library I refactored which called it  isHiddenOnLaunch so I shortened to hidden. I can change to minimized.
https://github.com/Teamwork/node-auto-launch/blob/master/src/AutoLaunchWindows.coffee

File: src/lib/autoLauncher.ts (hunk 7 above)
Comment (nazar-pc) [reply src/lib/autoLauncher.ts]: The changes implemented are not the changes I suggested. Please don't resolve things unless they are absolutely 100% resolved, it is annoying to go through resolved threads and unresolve them over and over again.

**Comment by jdheeter** [comment]
I was not able to test this working in Ubuntu because the app crashes on start #3 . Based on looking at the npm library from electron, it should work. I think safe to merge and check later when Linux build is not crashing.

File: src/lib/autoLauncher.ts (hunk 7 above)
Comment (jdheeter) [reply src/lib/autoLauncher.ts]: ah my mistake

Comment (nazar-pc) [reply src/lib/autoLauncher.ts]: It is more about contents than spacing actually

Comment (jdheeter) [reply src/lib/autoLauncher.ts]: ok updated

**Comment by jdheeter** [comment]
@nazar-pc I think ready to merge

**Review by nazar-pc — APPROVED** [review]

**Pull request merged**